app: support removing all units from the application #1111

lucasweiblen opened the pull request:
guys, could you please review this PR? if there is anything missing please say me. I'm just getting used to tsuru's codebase yet.

andrewsmedina commented:
LGTM.

fsouza commented on app/app.go:
```diff
@@ -398,10 +398,7 @@ func (app *App) RemoveUnits(n uint) error {
-		ReleaseApplicationLock(app.Name)
-		return stderr.New("Cannot remove all units from an app.")
-	} else if n > l {
+	} else if l := uint(len(app.Units())); n > l {
```
The change looks good, but I'd like to take this opportunity to rename this variable to "length". Using ``l`` as a variable name is confuse :-(

And I'm sorry, because I think I wrote this code in the first place. Do you mind just renaming this variable? After that, we can merge the pull request.

lucasweiblen replied:
no problems, man.. will fix this tonight

cezarsa commented:
Also, please notice that this PR solves half of the problem. After removing all units from an application, one should be able to call `tsuru unit-add` to bring the application back to life.
However, there's a check in the provisioner which doesn't allow adding units to applications with 0 units. This could be handled in another PR, but it has to go before everything works.

lucasweiblen commented:
hey @cezarsa, thanks for the feedback!
I will try to find this check in the provisioner in order to handle this second half of the problem.

lucasweiblen commented:
@cezarsa could you please confirm if this is the check you were talking about? thanks

https://github.com/tsuru/tsuru/blob/master/provision/docker/provisioner.go#L564:L566

cezarsa commented:
@lucasweiblen Yes, this check has to be changed. We cannot simply erase it because we still need to make sure the application have at least 1 deploy before adding new units. So we could change `length < 1` to `a.GetDeploys() == 0`. This might have a bit of an impact in this package's tests.

lucasweiblen commented:
thanks for the tip @cezarsa, I will have a look at this tonight.

lucasweiblen commented:
@cezarsa as you said , some tests(4 or 5) are now failing because the number of deploys is 0. To setup a deploy in the tests I would need to do something like this, right?

https://github.com/tsuru/tsuru/blob/master/provision/docker/provisioner_test.go#L104:L129

Is there a better way to do that?

cezarsa commented:
@lucasweiblen I think a more simple alternative is making the deploys attribute in FakeApp public, here:
https://github.com/tsuru/tsuru/blob/master/provision/provisiontest/fake_provisioner.go#L39
And in tests calling `AddUnits` test do something like:
```
app := provisiontest.NewFakeApp(...)
app.Deploys = 1
```

lucasweiblen commented:
@cezarsa much simpler approach! thank you very much.

I have changed AddUnits and fixed the broken tests. If anything is still missing please let me know!

fsouza commented:
@lucasweiblen the code looks good, can you rename the test "TestProvisionerAddUnitsWithoutContainers" to "TestProvisionerAddUnitsWithNoDeploys"?

Thanks!

lucasweiblen commented:
@fsouza done

fsouza commented:
LGTM.

Let's wait for @cezarsa :)

cezarsa commented:
Thanks a lot Lucas.
